### PR TITLE
[MRG] Initialize predictions to average value or class probabilities

### LINKED ISSUE #47: Initialize predictions to something better than just 0
Saving it here so we remember: instead of initializing `y_pred` to 0 we may want to do something more sensible like boosting from the average target, or even accept an `init` predictor.

### PR DESCRIPTION
**Pull request by NicolasHug**
Closes #47 

Predictions are now initialized to something a bit better than 0 for faster convergence. I based the initialization off of LGBM.

LGBM was using `boost_from_average=True` by default from the beginning, even in our `test_compare_lightgbm`. Weirdly enough, now that our code is more similar to theirs, I had to relax an accuracy check from 1% to 1.1% (only for 1 seed out of the 5).

### CODE REVIEW

File: pygbm/loss.py [hunk 1]
```diff
@@ -109,6 +116,13 @@ def __call__(self, y_true, raw_predictions, average=True):
         loss = np.logaddexp(0, raw_predictions) - y_true * raw_predictions
         return loss.mean() if average else loss
 
+    def get_init_prediction(self, y_train, n_trees_per_iteration):
+        proba_positive_class = np.mean(y_train)
+        eps = 1e-15
```
Comment (ogrisel) [inline pygbm/loss.py]: Maybe use `np.finfo(y_train.dtype).eps` instead.

File: pygbm/loss.py [hunk 2]
```diff
@@ -109,6 +116,13 @@ def __call__(self, y_true, raw_predictions, average=True):
...
+        proba_positive_class = np.clip(proba_positive_class, eps, 1 - eps)
+        # log(x / 1 - x) is the anti function of sigmoid
```
Comment (ogrisel) [inline pygbm/loss.py]: This is the link function of the Binomial predictive model.

File: pygbm/gradient_boosting.py [hunk 3]
```diff
@@ -156,18 +156,23 @@ def fit(self, X, y):
             print("Fitting gradient boosted rounds:")
 
         n_samples = X_binned_train.shape[0]
-        # values predicted by the trees. Used as-is in regression, and
-        # transformed into probas and / or classes for classification
+        self.init_predictions_ = self.loss_.get_init_prediction(
```
Comment (ogrisel) [inline pygbm/gradient_boosting.py]: I would rename this to `self.baseline_prediction_` (singular because it does not depend on the number of samples, even if for multiclass it's actually a n-dimensional single prediction :)

File: pygbm/loss.py [hunk 4]
```diff
@@ -156,6 +170,19 @@ def __call__(self, y_true, raw_predictions, average=True):
         return (logsumexp(raw_predictions, axis=1) -
                 (one_hot_true * raw_predictions).sum(axis=1))
 
+    def get_init_prediction(self, y_train, n_trees_per_iteration):
+        init_value = np.zeros(
+            shape=(1, n_trees_per_iteration),
+            dtype=np.float32
+        )
+        eps = 1e-15
```
Comment (ogrisel) [inline pygbm/loss.py]: `np.finfo(y_train.dtype).eps`

File: pygbm/gradient_boosting.py [hunk 5]
```diff
@@ -156,18 +156,23 @@ def fit(self, X, y):
...
+            y_train, self.n_trees_per_iteration_)
+        # raw_predictions are the values predicted by the trees for the
```
Comment (ogrisel) [inline pygbm/gradient_boosting.py]: the **accumulated** values predicted by the trees...

File: pygbm/gradient_boosting.py [hunk 6]
```diff
@@ -156,18 +156,23 @@ def fit(self, X, y):
...
+        # training data.
         raw_predictions = np.zeros(
             shape=(n_samples, self.n_trees_per_iteration_),
-            dtype=y_train.dtype
+            dtype=np.float32
```
Comment (ogrisel) [inline pygbm/gradient_boosting.py]: Maybe `dtype=self.init_predictions_.dtype`?

File: pygbm/gradient_boosting.py (hunk 3 above)
Comment (NicolasHug) [reply pygbm/gradient_boosting.py]: So `baseline_prediction_` or `singular_prediction_`?

I think I like `baseline` better

Comment (ogrisel) [reply pygbm/gradient_boosting.py]: baseline yes.

File: tests/test_loss.py [hunk 7]
```diff
@@ -152,3 +153,32 @@ def test_logsumexp():
...
+            assert_all_finite(init)
+
+    # check that the init values are constants
+    for loss_name, n_trees_per_iteration in (
+            ('binary_crossentropy', 1), ('categorical_crossentropy', 5)):
+        loss = _LOSSES[loss_name]()
+        y_true = rng.randint(0, n_trees_per_iteration + 1,
+                             size=1000).astype(np.float32)
+        init = loss.get_baseline_prediction(y_true, n_trees_per_iteration)
+        assert np.unique(init).size == n_trees_per_iteration
```
Comment (ogrisel) [inline tests/test_loss.py]: Could you please split this test in 3 tests: one for each loss and actually test that the baseline prediction matches the class frequency ratios for the classification losses.

Please also add assertions on the expected shape of `loss.get_baseline_prediction(y_train, n_trees_per_iteration)` in each case.

File: tests/test_loss.py [hunk 8]
```diff
@@ -152,3 +153,32 @@ def test_logsumexp():
...
+            loss = _LOSSES[loss_name]()
+            init = loss.get_baseline_prediction(y_true, n_trees_per_iteration)
+            assert_all_finite(init)
+
+    # check that the init values are constants
+    for loss_name, n_trees_per_iteration in (
+            ('binary_crossentropy', 1), ('categorical_crossentropy', 5)):
+        loss = _LOSSES[loss_name]()
+        y_true = rng.randint(0, n_trees_per_iteration + 1,
+                             size=1000).astype(np.float32)
```
Comment (ogrisel) [inline tests/test_loss.py]: This should be named `y_train`, right?

File: pygbm/loss.py [hunk 9]
```diff
@@ -156,6 +185,19 @@ def __call__(self, y_true, raw_predictions, average=True):
         return (logsumexp(raw_predictions, axis=1) -
                 (one_hot_true * raw_predictions).sum(axis=1))
 
+    def get_baseline_prediction(self, y_train, n_trees_per_iteration):
```
Comment (ogrisel) [inline pygbm/loss.py]: I would rename `n_trees_per_iteration` as `prediction_dim` in the context of the loss function classes.

File: tests/test_loss.py (hunk 7 above)
Comment (NicolasHug) [reply tests/test_loss.py]: >test that the baseline prediction matches the class frequency ratios for the classification losses.

You mean check that the values are exactly as expected (log(1 / 1-p))? I'm not sure how to do that without simply copy pasting the code from the loss, but then this test doesn't check much.

File: pygbm/loss.py (hunk 9 above)
Comment (NicolasHug) [reply pygbm/loss.py]: I'll do it but I'm not sure it's always better. For example in `_update_gradients_hessians_categorical_crossentropy`, the update is semantically related to a tree, not to a 'dimension'.

File: tests/test_loss.py (hunk 7 above)
Comment (ogrisel) [reply tests/test_loss.py]: Yes, consider this simple test as a non regression test and a way to clarify what is the intent of baseline prediction.

File: pygbm/loss.py (hunk 9 above)
Comment (ogrisel) [reply pygbm/loss.py]: That's a good point. I was thinking that at some point we might want to try to implement multioutput regression trees such that we can have a categorical loss with one 1 tree per iteration. Anyways, we might rework that later.

Comment (ogrisel) [reply pygbm/loss.py]: My point was to rename this local variable to `prediction_dim` only in the context of the loss function but keep the `n_trees_per_iteration` in the context of gradient boosting iteration.

But I think it's also ok to use the name `raw_prediction_dim` or `prediction_dim` everywhere.

File: tests/test_loss.py [hunk 10]
```diff
@@ -152,3 +153,33 @@ def test_logsumexp():
...
+    assert baseline_prediction.shape == tuple()  # float
+    assert_almost_equal(baseline_prediction, y_train.mean())
+
+
+def test_baseline_binary_crossentropy():
+    loss = _LOSSES['binary_crossentropy']()
+    for y_train in (np.zeros(shape=100), np.ones(shape=100)):
+        y_train = y_train.astype(np.float32)
+        baseline_prediction = loss.get_baseline_prediction(y_train, 1)
+        assert baseline_prediction.shape == tuple()  # float
```
Comment (ogrisel) [inline tests/test_loss.py]: nitpick "scalar" would be a better description than "float".

File: tests/test_loss.py [hunk 11]
```diff
@@ -152,3 +153,33 @@ def test_logsumexp():
...
+
+def test_baseline_categorical_crossentropy():
+    prediction_dim = 4
+    loss = _LOSSES['categorical_crossentropy']()
+    for y_train in (np.zeros(shape=100), np.ones(shape=100)):
+        y_train = y_train.astype(np.float32)
+        baseline_prediction = loss.get_baseline_prediction(y_train,
+                                                           prediction_dim)
+        assert baseline_prediction.shape == (1, prediction_dim)
+        assert_all_finite(baseline_prediction)
```
Comment (ogrisel) [inline tests/test_loss.py]: For those loss test, one could check that applying `_inverse_link_function(base_line_prediction)` approximately recovers `y_train[0]`.

File: tests/test_loss.py (hunk 7 above)
Comment (ogrisel) [reply tests/test_loss.py]: Or even consider this test as fine grained specification/documentation from the API consumer point of view.

**Comment by ogrisel** [comment]
Thanks!
